File: backend/src/services/integration.service.ts
import type { RFIDService } from './rfid.service'
import type { SonosService } from './sonos.service'
import type { CardMappingService } from './card-mapping.service'
import type { WSServer } from '../websocket'
import type { RFIDEvent } from '../types/rfid.types'
import type { SonosState } from '../types/sonos.types'
import type { IncomingMessage } from '../types/websocket.types'
import type { WebSocket } from 'ws'

export class IntegrationService {
  constructor(
    private rfid: RFIDService,
    private sonos: SonosService,
    private cardMappings: CardMappingService,
    private ws: WSServer,
  ) {
    this.setupEventHandlers()
  }

  private setupEventHandlers(): void {
    // RFID events → Sonos actions + Frontend notifications
    this.rfid.on('event', (event: RFIDEvent) => {
      this.handleRFIDEvent(event)
    })

    // Sonos state changes → Frontend updates
    this.sonos.on('state', (state: SonosState) => {
      this.ws.broadcast({
        type: 'sonos_state',
        ...state,
      })
    })

    // Frontend messages → RFID/Sonos actions
    this.ws.onMessage((message: IncomingMessage, client: WebSocket) => {
      this.handleFrontendMessage(message, client)
    })

    console.log('Integration service event handlers setup complete')
  }

  private async handleRFIDEvent(event: RFIDEvent): Promise<void> {
    if (event.type === 'card_read') {
      console.log(`Card read: ${event.cardId}`)

      // Look up card mapping
      const mapping = this.cardMappings.getMapping(event.cardId)

      if (!mapping) {
        console.warn(`No mapping found for card ${event.cardId}`)
        this.ws.broadcast({
          type: 'card_read',
          cardId: event.cardId,
          data: null,
          timestamp: event.timestamp,
          error: 'Card not registered',
        })
        return
      }

      console.log(`Card ${event.cardId} mapped to ${mapping.type} action`)

      // Broadcast card read to frontend
      this.ws.broadcast({
        type: 'card_read',
        cardId: event.cardId,
        data: mapping.data,
        actionType: mapping.type,
        name: mapping.name,
        timestamp: event.timestamp,
      })

      // Execute action based on type
      await this.executeCardAction(mapping)
    } else if (event.type === 'write_complete') {
      // Forward write completion to frontend
      this.ws.broadcast({
        type: 'write_complete',
        success: event.success,
        error: event.error,
      })
    } else if (event.type === 'error') {
      // Forward RFID errors to frontend
      this.ws.broadcast({
        type: 'error',
        source: 'rfid',
        message: event.message,
      })
    }
  }

  private async executeCardAction(
    mapping: import('../types/card-mapping.types').CardMapping,
  ): Promise<void> {
    try {
      switch (mapping.type) {
        case 'sonos':
          // Play Sonos content
          if (typeof mapping.data === 'string') {
            // Check if it's a URI or a favorite title
            if (this.isUri(mapping.data)) {
              await this.sonos.playUrl(mapping.data)
            } else {
              // Treat as favorite title
              await this.sonos.playFavoriteByTitle(mapping.data)
            }
          } else {
            console.error('Invalid Sonos data format - expected string')
          }
          break

        // Future integrations will go here:
        // case 'lights':
        //   await this.lights.setScene(mapping.data)
        //   break

        default:
          console.warn(`Unknown action type: ${mapping.type}`)
      }
    } catch (error) {
      console.error(`Failed to execute ${mapping.type} action:`, error)
      this.ws.broadcast({
        type: 'error',
        source: mapping.type,
        message: error instanceof Error ? error.message : String(error),
      })
    }
  }

  private isUri(str: string): boolean {
    // Check if string looks like a URI (has a scheme)
    return (
      str.startsWith('x-') ||
      str.startsWith('http://') ||
      str.startsWith('https://') ||
      str.startsWith('spotify:') ||
      str.includes('://')
    )
  }

  private async handleFrontendMessage(
    message: IncomingMessage,
    client: WebSocket,
  ): Promise<void> {
    try {
      switch (message.type) {
        case 'sonos_play_url':
          await this.sonos.playUrl(message.url)
          break

        case 'sonos_play':
          await this.sonos.play()
          break

        case 'sonos_pause':
          await this.sonos.pause()
          break

        case 'sonos_next':
          await this.sonos.next()
          break

        case 'sonos_previous':
          await this.sonos.previous()
          break

        case 'sonos_volume':
          await this.sonos.setVolume(message.volume)
          break

        case 'write_request':
          this.rfid.sendCommand({
            type: 'write_request',
            data: message.data,
          })
          break

        default:
          console.warn('Unknown message type:', (message as any).type)
      }
    } catch (error) {
      console.error('Error handling frontend message:', error)
      this.ws.send(client, {
        type: 'error',
        source: 'system',
        message: error instanceof Error ? error.message : String(error),
      })
    }
  }

}

File: backend/services/integration_service.py
import logging
import time
from typing import Dict, Any
from fastapi import WebSocket

from .rfid_service import RFIDService
from .sonos_service import SonosService, SonosState
from .card_mapping_service import CardMappingService
from .websocket_service import WebSocketManager

logger = logging.getLogger(__name__)


class IntegrationService:
    """
    Orchestration layer connecting all services.
    Routes events between RFID, Sonos, Card Mappings, and WebSocket.
    """

    def __init__(
        self,
        rfid: RFIDService,
        sonos: SonosService,
        card_mappings: CardMappingService,
        ws_manager: WebSocketManager,
    ):
        self.rfid = rfid
        self.sonos = sonos
        self.card_mappings = card_mappings
        self.ws_manager = ws_manager

    async def setup(self) -> None:
        """Setup event handlers and start services"""
        # Start RFID polling with card read callback
        await self.rfid.start_polling(self.handle_card_read)

        # Start Sonos state polling with state update callback
        await self.sonos.start_polling(self.handle_sonos_state)

        logger.info("Integration service event handlers setup complete")

    async def handle_card_read(self, card_id: str, data: str) -> None:
        """
        Handle RFID card read event.
        Looks up card mapping and executes the configured action.
        """
        logger.info(f"Card read: {card_id}")

        # Look up card mapping
        mapping = self.card_mappings.get_mapping(card_id)

        if not mapping:
            logger.warning(f"No mapping found for card {card_id}")
            await self.ws_manager.broadcast(
                {
                    "type": "card_read",
                    "cardId": card_id,
                    "data": None,
                    "timestamp": time.time(),
                    "error": "Card not registered",
                }
            )
            return

        logger.info(f"Card {card_id} mapped to {mapping.type} action")

        # Broadcast card read to frontend
        await self.ws_manager.broadcast(
            {
                "type": "card_read",
                "cardId": card_id,
                "data": mapping.data,
                "actionType": mapping.type,
                "name": mapping.name,
                "timestamp": time.time(),
            }
        )

        # Execute action based on type
        await self.execute_card_action(mapping)

    async def execute_card_action(self, mapping: Any) -> None:
        """Execute the action configured for a card"""
        try:
            if mapping.type == "sonos":
                # Play Sonos content
                if isinstance(mapping.data, str):
                    # Check if it's a URI or a favorite title
                    if self._is_uri(mapping.data):
                        await self.sonos.play_url(mapping.data)
                    else:
                        # Treat as favorite title
                        await self.sonos.play_favorite_by_title(mapping.data)
                else:
                    logger.error("Invalid Sonos data format - expected string")

            # Future integrations go here:
            # elif mapping.type == "lights":
            #     await self.lights.set_scene(mapping.data)

            else:
                logger.warning(f"Unknown action type: {mapping.type}")

        except Exception as error:
            logger.error(f"Failed to execute {mapping.type} action: {error}")
            await self.ws_manager.broadcast(
                {
                    "type": "error",
                    "source": mapping.type,
                    "message": str(error),
                }
            )

    def _is_uri(self, string: str) -> bool:
        """Check if string looks like a URI (has a scheme)"""
        return (
            string.startswith("x-")
            or string.startswith("http://")
            or string.startswith("https://")
            or string.startswith("spotify:")
            or "://" in string
        )

    async def handle_sonos_state(self, state: SonosState) -> None:
        """
        Handle Sonos state update.
        Broadcasts state to all WebSocket clients.
        """
        # Convert to dict for JSON serialization
        message = {
            "type": "sonos_state",
            "playbackState": state.playback_state,
            "isPlaying": state.is_playing,
            "volume": state.volume,
            "speakerName": state.speaker_name,
            "currentTrack": None,
        }

        if state.current_track:
            message["currentTrack"] = {
                "title": state.current_track.title,
                "artist": state.current_track.artist,
                "album": state.current_track.album,
                "albumArtUri": state.current_track.album_art_uri,
                "duration": state.current_track.duration,
                "position": state.current_track.position,
            }

        await self.ws_manager.broadcast(message)

    async def handle_websocket_message(
        self, message: Dict[str, Any], websocket: WebSocket
    ) -> None:
        """
        Handle incoming WebSocket message from frontend.
        Routes to appropriate service.
        """
        try:
            msg_type = message.get("type")

            if msg_type == "sonos_play_url":
                await self.sonos.play_url(message["url"])

            elif msg_type == "sonos_play":
                await self.sonos.play()

            elif msg_type == "sonos_pause":
                await self.sonos.pause()

            elif msg_type == "sonos_next":
                await self.sonos.next()

            elif msg_type == "sonos_previous":
                await self.sonos.previous()

            elif msg_type == "sonos_volume":
                await self.sonos.set_volume(message["volume"])

            elif msg_type == "write_request":
                success = await self.rfid.write_card(message["data"])
                await self.ws_manager.send_personal(
                    {
                        "type": "write_complete",
                        "success": success,
                    },
                    websocket,
                )

            else:
                logger.warning(f"Unknown message type: {msg_type}")

        except Exception as error:
            logger.error(f"Error handling frontend message: {error}")
            await self.ws_manager.send_personal(
                {
                    "type": "error",
                    "source": "system",
                    "message": str(error),
                },
                websocket,
            )

    async def shutdown(self) -> None:
        """Cleanup on shutdown"""
        logger.info("Shutting down integration service...")
        self.rfid.stop_polling()
        self.sonos.stop_polling()
        logger.info("Integration service shutdown complete")
